lib/bolt/plugin/azure.rb

```diff
@@ -0,0 +1,199 @@
+            'name' => instance.dig('properties', 'dnsSettings', 'fqdn'),
+            'uri' => instance.dig('properties', 'ipAddress')
+          }.compact
+
+          target unless target.empty?
+        end.compact
+      end
+
+      # Hash of required credentials for authorizing with the Azure REST API
+      # These values can be set in 3 locations - inventory config, Bolt config, environment variables
+      # TODO: Add support for reading credentials from .ini file?
+      def credentials(opts)
+        creds = {
+          'tenant_id' => (opts['tenant_id'] || config['tenant_id'] || ENV['AZURE_TENANT_ID']),
+          'client_id' => (opts['client_id'] || config['client_id'] || ENV['AZURE_CLIENT_ID']),
+          'client_secret' => (opts['client_secret'] || config['client_secret'] || ENV['AZURE_CLIENT_SECRET']),
+          'subscription_id' => (opts['subscription_id'] || config['subscription_id'] || ENV['AZURE_SUBSCRIPTION_ID'])
+        }
+
+        # All credentials must be set, otherwise there's no point continuing
+        if creds.values.include? nil
+          keys = creds.select { |_, v| v.nil? }.keys
+          raise Bolt::ValidationError, "Missing required credentials: #{keys}"
+        end
+
+        creds
+      end
+
+      # Requests for VMs and scale sets are on a per-subscription basis
+      # You can also request VMs by a specific resource group
+      # Scale sets are always requested by resource group and scale set name
+      #
+      # Since each request only returns up to 1,000 results, requests will continue to be
+      # sent until there is no longer a nextLink token in the result set
+      def instances(token, creds, opts)
+        url = if opts['resource_group']
+                if opts['scale_set']
+                  "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                  "resourceGroups/#{opts['resource_group']}/providers/Microsoft.Compute/" \
+                  "virtualMachineScaleSets/#{opts['scale_set']}/" \
+                  "publicipaddresses?api-version=2017-03-30"
+                else
+                  "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                  "resourceGroups/#{opts['resource_group']}/providers/Microsoft.Network/" \
+                  "publicIPAddresses?api-version=2019-06-01"
+                end
+              else
+                "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                "providers/Microsoft.Network/publicIPAddresses?api-version=2019-06-01"
+              end
+
+        header = {
+          'Authorization' => "#{token['token_type']} #{token['access_token']}"
+        }
+
+        instances = []
+
+        while url do
+          # Update the URI and make the next request
+          uri = URI.parse(url)
+          result = request(:Get, uri, nil, header)
+
+          # Add the VMs to the list of instances
+          instances << result['value']
```

Comment:
Sorry! I was mistaken. This should be `concat` (because `result['value']` is an array). Then the `flatten` call is unnecessary.